designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# Moving to Asynchronous API
+
+## Summary
+
+This RFC adds a new class `ESLint` that has asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- We have the functionality that we cannot support with synchronous API. For example, ESLint verifies files in parallel, formatters print progress state, or formatters print results in streaming.
+- Dynamic `import()` has arrived at Stage 4. The dynamic loading of ES modules requires asynchronous API. Migrating to asynchronous API opens up doors to write plugins/configs with ES modules.
+- The name of `CLIEngine`, our primary API, has caused confusion to the community. People try `Linter` class at first because the name sounds the main class of ESLint, then they notice it doesn't work as expected. We have a lot of issues that say "please use `CLIEngine` instead." The name of the new class, `ESLint`, is our primary API clearly.
+
+## Detailed Design
+
+### ■ Add new `ESLint` class
+
+This RFC adds a new class `ESLint`. It has almost the same methods as `CLIEngine`, but the return value of some methods are different.
+
+So, for now, `ESLint` class will be a tiny wrapper of `CLIEngine` that modifies the type of the return values.
+
+#### § Constructor
+
+The constructor has the most same arguments as `CLIEngine`, but there are small differences.
+
+- It throws fatal errors if the options contain unknown properties or an option is invalid type ([eslint/eslint#10272](https://github.com/eslint/eslint/issues/10272)).
+- It disallows the deprecated `cacheFile` option.
+- It has a new `pluginImplementations` option as the successor of `addPlugin()` method. This is an object that keys are plugin IDs and each value is the plugin object.
```

Comment:
> - Is it because they are breaking changes and that it would be beneficial to package all breaking changes together when it comes to creating a new class like ESLint so that there won't be a need for another breaking change close after it?
> - Does that also mean that you intend these fixes to purely be made in the new ESLint class and none of them in the existing CLIEngine?

Exactly.

> - Are there references to any existing issues for the two unreferenced differences? Like the cacheFile one?

- I didn't search the issue of `cacheFile`, but [the document](https://eslint.org/docs/developer-guide/nodejs-api#cliengine) says that it's deprecated.
- About `addPlugin`, I have remembered it by a recent issue https://github.com/eslint/eslint/issues/12425 then updated this RFC.

> - Does that mean that addPlugin() gets removed or deprecated?

A later section explains it, ESLint class doesn't have the `addPlugin()` method. the `CLIEngine#addPlugin()` method is as-is, but `CLIEngine` is deprecated.